lib/portage/tests/dbapi/test_bintree.py

```diff
@@ -0,0 +1,155 @@
+            "invalids",
+            "settings",
+            "_pkg_paths",
+            "_populating",
+            "_all_directory",
+            "_pkgindex_version",
+            "_pkgindex_hashes",
+            "_pkgindex_file",
+            "_pkgindex_keys",
+            "_pkgindex_aux_keys",
+            "_pkgindex_use_evaluated_keys",
+            "_pkgindex_header",
+            "_pkgindex_header_keys",
+            "_pkgindex_default_pkg_data",
+            "_pkgindex_inherited_keys",
+            "_pkgindex_default_header_data",
+            "_pkgindex_translated_keys",
+            "_pkgindex_allowed_pkg_keys",
+        }
+        no_multi_instance_settings = MagicMock()
+        no_multi_instance_settings.features = ""
+        no_multi_instance_bt = binarytree(
+            pkgdir="/tmp", settings=no_multi_instance_settings
+        )
+        multi_instance_settings = MagicMock()
+        multi_instance_settings.features = "binpkg-multi-instance"
+        multi_instance_bt = binarytree(pkgdir="/tmp", settings=multi_instance_settings)
+        for attr in required_attrs_no_multi_instance:
+            getattr(no_multi_instance_bt, attr)
+            getattr(multi_instance_bt, attr)
+        # The next attribute is the difference between multi instance
+        # and no multi instance:
+        getattr(multi_instance_bt, "_allocate_filename")
+
+    @patch("portage.dbapi.bintree.binarytree._populate_local")
+    def test_populate_without_updates_repos_nor_getbinspkgs(self, ppopulate_local):
+        bt = binarytree(pkgdir="/tmp", settings=MagicMock())
+        ppopulate_local.return_value = {}
+        bt.populate()
+        ppopulate_local.assert_called_once_with(reindex=True)
+        self.assertFalse(bt._populating)
+        self.assertTrue(bt.populated)
+
+    @patch("portage.dbapi.bintree.binarytree._populate_local")
+    def test_populate_calls_twice_populate_local_if_updates(self, ppopulate_local):
+        bt = binarytree(pkgdir="/tmp", settings=MagicMock())
+        bt.populate()
+        self.assertIn(call(reindex=True), ppopulate_local.mock_calls)
+        self.assertIn(call(), ppopulate_local.mock_calls)
+        self.assertEqual(ppopulate_local.call_count, 2)
+
+    @patch("portage.dbapi.bintree.binarytree._populate_additional")
+    @patch("portage.dbapi.bintree.binarytree._populate_local")
+    def test_populate_with_repos(self, ppopulate_local, ppopulate_additional):
+        repos = ("one", "two")
+        bt = binarytree(pkgdir="/tmp", settings=MagicMock())
+        bt.populate(add_repos=repos)
+        ppopulate_additional.assert_called_once_with(repos)
+
+    @patch("portage.dbapi.bintree.BinRepoConfigLoader")
+    @patch("portage.dbapi.bintree.binarytree._populate_remote")
+    @patch("portage.dbapi.bintree.binarytree._populate_local")
+    def test_populate_with_getbinpkgs(
+        self, ppopulate_local, ppopulate_remote, pBinRepoConfigLoader
+    ):
+        refresh = "something"
```

Comment:
I'm confused by this. Why are you passing a string in the getbinpkg_refresh kwarg?

Reply:
The idea of this test is to make sure that if `populate` is called with `getbinpkgs=True`, then `_populate_remote` is called, and it is *blindly* given whatever is passed in the `getbinpkg_refresh` argument of `populate`.
This test is also telling us that there is no validation done by `populate` and the semantics of the `getbinpkg_refresh` parameter (and all the responsibility to handle it) is translated to `_populate_remote`.

I chose a meaningless string to make clear the point that for `populate` it is irrelevant what is the value of that parameter.

Do you think that this is wrong or unnecessary? Do you have a better idea?

Reply:
Ah, thanks for explaining.